### PR TITLE
Refactor operation tracker ctor and introduce custom percentiles

### PR DESCRIPTION
**Pull request by jsjtzyy**
1. refactor ctor of both simple and adaptive operation tracker by
passing the router config and operation class. This allows operation
tracker itself to populate its parameters based on type of operation
class.
2. introduce custom percentiles of latency Histogram in adaptive
operation tracker. The percentiles are configurable and on-demand. User
can define arbitrary quantile like 91th percentile to better eveluate
the latency distribution.

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/GetOperation.java [hunk 1]
```diff
@@ -237,23 +237,19 @@ protected GetRequest createGetRequest(BlobId blobId, MessageFormatFlags flag, Ge
   /**
    * Gets an {@link OperationTracker} based on the config and {@code partitionId}.
    * @param partitionId the {@link PartitionId} for which a tracker is required.
+   * @param operationClass
```
Comment (zzmao) [inline ambry-router/src/main/java/com.github.ambry.router/GetOperation.java]: comment

File: ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java [hunk 2]
```diff
@@ -289,5 +304,9 @@ public RouterConfig(VerifiableProperties verifiableProperties) {
         verifiableProperties.getIntInRange("router.ttl.update.success.target", 2, 1, Integer.MAX_VALUE);
     routerUseGetBlobOperationForBlobInfo =
         verifiableProperties.getBoolean("router.use.get.blob.operation.for.blob.info", false);
+    List<String> customPercentiles =
+        Utils.splitString(verifiableProperties.getString("router.operation.tracker.custom.percentiles", ""), ",");
+    routerOperationTrackerCustomPercentiles =
+        Collections.unmodifiableList(customPercentiles.stream().map(Double::valueOf).collect(Collectors.toList()));
```
[the remark this answers is not in the document]
Comment (zzmao) [reply ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java]: got you. It's OK.

**Comment by jsjtzyy** [comment]
./gradlew build && ./gradlew test succeeded. Reminder to review, thanks.

File: ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java [hunk 3]
```diff
@@ -72,25 +74,43 @@
...
    */
-  SimpleOperationTracker(String datacenterName, PartitionId partitionId, boolean crossColoEnabled,
-      String originatingDcName, boolean includeNonOriginatingDcReplicas, int replicasRequired, int successTarget,
-      int parallelism, boolean shuffleReplicas) {
+  SimpleOperationTracker(RouterConfig routerConfig, Class operationClass, PartitionId partitionId,
+      String originatingDcName, boolean shuffleReplicas) {
+    // populate tracker parameters based on operation type
+    boolean crossColoEnabled = false;
+    boolean includeNonOriginatingDcReplicas = true;
+    int replicasRequired = Integer.MAX_VALUE;
```
Comment (cgtz) [inline ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java]: how is this local variable used?

File: ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java [hunk 4]
```diff
@@ -72,25 +74,43 @@
...
+    if (GetOperation.class.isAssignableFrom(operationClass)) {
+      successTarget = routerConfig.routerGetSuccessTarget;
+      parallelism = routerConfig.routerGetRequestParallelism;
+      crossColoEnabled = routerConfig.routerGetCrossDcEnabled;
+      includeNonOriginatingDcReplicas = routerConfig.routerGetIncludeNonOriginatingDcReplicas;
+      replicasRequired = routerConfig.routerGetReplicasRequired;
+    } else if (operationClass == PutOperation.class) {
```
Comment (cgtz) [inline ambry-router/src/main/java/com.github.ambry.router/SimpleOperationTracker.java]: One other option is to make an enum to represent the operation tracker types (i.e. GET, PUT, DELETE, TTL_UPDATE). This is a little more explicit and resilient to new classes/name changes.

File: ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java [hunk 5]
```diff
@@ -232,6 +238,15 @@
...
   public final boolean routerUseGetBlobOperationForBlobInfo;
 
+  /**
+   * The custom percentiles of Histogram in operation tracker to be reported. This allows router to emit metrics of
+   * arbitrary percentiles (i.e. 97th, 93th etc). An example of this config is "0.91,0.93,0.97"(comma separated), each
+   * value should fall in {@code [0..1]}.
+   */
+  @Config("router.operation.tracker.custom.percentiles")
+  @Default("")
+  public final List<Double> routerOperationTrackerCustomPercentiles;
```
Comment (cgtz) [inline ambry-api/src/main/java/com.github.ambry/config/RouterConfig.java]: how will these custom percentiles be used by operation tracker in the future? Right now, they are only registered in the `MetricRegistry` for viewing.

**Comment by lightningrob** [comment]
@zzmao or @cgtz Can one of you merge this?

**Comment by zzmao** [comment]
Good to merge? @cgtz

**Pull request merged**